File: dolphin-launcher-v1/app/src/main/java/com/dolphin/launcher/v1/BydPermissionContext.java
package com.dolphin.launcher.v1;

import android.content.Context;
import android.content.ContextWrapper;
import android.content.pm.PackageManager;

/**
 * Narrow compatibility wrapper used only for BYD framework getInstance(Context).
 * It mirrors the legacy real-car-tested client-side BYDAUTO permission behavior
 * while leaving every unrelated Android permission check untouched.
 */
public final class BydPermissionContext extends ContextWrapper {
    private static final String PREFIX = "android.permission.BYDAUTO_";

    private BydPermissionContext(Context base) {
        super(base.getApplicationContext());
    }

    public static Context wrap(Context context) {
        if (context instanceof BydPermissionContext) return context;
        return new BydPermissionContext(context);
    }

    private static boolean isByd(String permission) {
        return permission != null && permission.startsWith(PREFIX);
    }

    @Override
    public void enforceCallingOrSelfPermission(String permission, String message) {
        if (isByd(permission)) return;
        super.enforceCallingOrSelfPermission(permission, message);
    }

    @Override
    public int checkCallingOrSelfPermission(String permission) {
        return isByd(permission) ? PackageManager.PERMISSION_GRANTED
                : super.checkCallingOrSelfPermission(permission);
    }

    @Override
    public void enforcePermission(String permission, int pid, int uid, String message) {
        if (isByd(permission)) return;
        super.enforcePermission(permission, pid, uid, message);
    }

    @Override
    public int checkPermission(String permission, int pid, int uid) {
        return isByd(permission) ? PackageManager.PERMISSION_GRANTED
                : super.checkPermission(permission, pid, uid);
    }
}

File: nextapp/src/main/java/com/byd/dolphin/autoassistant/next/core/BydPermissionContext.kt
package com.byd.dolphin.autoassistant.next.core

import android.content.Context
import android.content.ContextWrapper
import android.content.pm.PackageManager

class BydPermissionContext private constructor(base: Context) : ContextWrapper(base.applicationContext) {
    override fun checkCallingOrSelfPermission(permission: String): Int =
        if (permission.startsWith(PREFIX)) PackageManager.PERMISSION_GRANTED
        else super.checkCallingOrSelfPermission(permission)

    override fun checkPermission(permission: String, pid: Int, uid: Int): Int =
        if (permission.startsWith(PREFIX)) PackageManager.PERMISSION_GRANTED
        else super.checkPermission(permission, pid, uid)

    override fun enforceCallingOrSelfPermission(permission: String, message: String?) {
        if (!permission.startsWith(PREFIX)) super.enforceCallingOrSelfPermission(permission, message)
    }

    override fun enforcePermission(permission: String, pid: Int, uid: Int, message: String?) {
        if (!permission.startsWith(PREFIX)) super.enforcePermission(permission, pid, uid, message)
    }

    companion object {
        private const val PREFIX = "android.permission.BYDAUTO_"
        fun wrap(context: Context): Context =
            if (context is BydPermissionContext) context else BydPermissionContext(context)
    }
}
